kitsune/wiki/views.py

```diff
@@ -483,94 +483,123 @@ def edit_document(request, document_slug, revision_id=None):
-                        )
-                disclose_description = True
-            else:
-                raise PermissionDenied
-        elif which_form == "rev":
-            if doc.allows(user, "create_revision"):
-                rev_form = RevisionForm(request.POST)
-                rev_form.instance.document = doc  # for rev_form.clean()
-                if rev_form.is_valid():
-                    _save_rev_and_notify(rev_form, user, doc, base_rev=rev)
-                    if "notify-future-changes" in request.POST:
-                        EditDocumentEvent.notify(request.user, doc)
+    if doc.locale != request.LANGUAGE_CODE:
+        # We've fallen back to the parent, since no visible translation existed.
+        url = reverse("wiki.translate", locale=request.LANGUAGE_CODE, args=[document_slug])
+        return HttpResponseRedirect(url)
 
-                    return HttpResponseRedirect(
-                        reverse("wiki.document_revisions", args=[document_slug])
-                    )
+    can_edit_needs_change = doc.allows(user, "edit_needs_change")
+    can_archive = doc.allows(user, "archive")
+
+    # If this document has a parent, then the edit is handled by the
+    # translate view. Pass it on.
+    if doc.parent:
+        return translate(request, doc.parent.slug, revision_id)
+    if revision_id:
+        rev = get_object_or_404(Revision, pk=revision_id, document=doc)
+    else:
+        rev = doc.current_revision or doc.revisions.order_by("-created", "-id")[0]
+
+    if request.method == "POST":  # POST
+        _document_lock_clear(doc.id, user.username)
+
+        post_data = request.POST.copy()
+        post_data.update({"locale": request.LANGUAGE_CODE})
+
+        topics = []
+        for t in post_data.getlist("topics"):
+            topics.append(int(t))
+        post_data.setlist("topics", topics)
+
+        doc_form = DocumentForm(
+            post_data,
+            instance=doc,
+            can_archive=can_archive,
+            can_edit_needs_change=can_edit_needs_change,
+        )
+        if doc_form.is_valid():
+            # Get the possibly new slug for the imminent redirection:
+            try:
+                doc = doc_form.save(None)
+            except (TitleCollision, SlugCollision) as e:
+                # TODO: .add_error() when we upgrade to Django 1.7
+                errors = doc_form._errors.setdefault("title", ErrorList())
+                message = "The {type} you selected is already in use."
+                message = message.format(type="title" if isinstance(e, TitleCollision) else "slug")
+                errors.append(_(message))
             else:
-                raise PermissionDenied
+                # Do we need to rebuild the KB?
+                _maybe_schedule_rebuild(doc_form)
+                return HttpResponseRedirect(reverse("wiki.document", args=[doc.slug]))
+    else:  # GET
+        # We can't drop this ELSE because we need to account for bad POSTs
```

Comment:
If the form is not valid, won't the form's clean method raise the errors before reaching this point?

Reply:
If `is_valid()` fails we want to render the same form submission, with the errors. So if `is_valid()` fails during a `POST`, execution continues to the end. We won't create a new instance of DocumentForm (which we do for a GET) but instead use the same `doc_form`. So you get the errors with your form data vs. new form data.